File: backend/app/schemas/agent_control.py
"""Pydantic schemas for the agent-control router (#235).

Wire format for the control-tower view: `app.models.agent_discovery.DiscoveredAgent`
enriched with derived health + lifecycle, and the persisted overrides from
`app.models.agent_control.AgentControlState`. See
`services/agent_control_service.py` for how the two are combined.
"""

from __future__ import annotations

import enum
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.agent_control import AgentControlAction
from app.models.agent_discovery import AgentCloudProvider, AgentDiscoveryStatus


class AgentHealth(str, enum.Enum):
    """Derived from `DiscoveredAgent.last_seen_at` recency — not persisted."""

    healthy = "healthy"
    degraded = "degraded"
    unhealthy = "unhealthy"


class AgentLifecycle(str, enum.Enum):
    """Discovery status, overridden by a persisted pause/quarantine action."""

    unmonitored = "unmonitored"
    provisioning = "provisioning"
    active = "active"
    paused = "paused"
    quarantined = "quarantined"


class AgentControlStateResponse(BaseModel):
    """One governed agent's control-tower view: discovery identity + derived/persisted state."""

    id: str
    provider: AgentCloudProvider
    name: str
    registry: str
    discovery_status: AgentDiscoveryStatus
    health: AgentHealth
    lifecycle: AgentLifecycle
    guardrail_enabled: bool
    last_action: AgentControlAction | None = None
    last_action_at: datetime | None = None
    last_seen_at: datetime


class AgentControlListResponse(BaseModel):
    """All governed agents for the tenant, control-tower enriched."""

    agents: list[AgentControlStateResponse] = Field(default_factory=list)


class AgentControlActionRequest(BaseModel):
    """Body of `POST /agent-control/{id}/action`."""

    action: AgentControlAction

File: frontend/src/lib/aispm/agent-control.ts
'use client';

/**
 * Agent control tower API client and types (M1, issue #236).
 *
 * Mirrors the backend Pydantic schemas in `app/schemas/agent_control.py`
 * and the `/api/v1/agent-control` router landed in #235: health and
 * lifecycle are derived server-side from discovery data + persisted
 * overrides, this module just wires the read + action calls.
 *
 * `AgentCloudProvider`/`AgentDiscoveryStatus` are duplicated from
 * `lib/aispm/agent-discovery.ts` (#234, not yet merged as of this issue) —
 * once that module lands on main, re-export from there instead.
 */

import { api } from '../api';

// ──────────────────────────────────────────────────────────────────────
// Enums (string unions mirroring the backend enums)
// ──────────────────────────────────────────────────────────────────────

export type AgentCloudProvider = 'aws_bedrock_agentcore' | 'azure_ai_foundry' | 'gcp_knowledge_catalog';

export type AgentDiscoveryStatus = 'shadow' | 'pending' | 'approved' | 'dismissed';

export type AgentHealth = 'healthy' | 'degraded' | 'unhealthy';

export type AgentLifecycle = 'unmonitored' | 'provisioning' | 'active' | 'paused' | 'quarantined';

export type AgentControlAction = 'pause' | 'quarantine' | 'toggle-guardrail';

// ──────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────

export interface AgentControlState {
  id: string;
  provider: AgentCloudProvider;
  name: string;
  registry: string;
  discovery_status: AgentDiscoveryStatus;
  health: AgentHealth;
  lifecycle: AgentLifecycle;
  guardrail_enabled: boolean;
  last_action: AgentControlAction | null;
  last_action_at: string | null;
  last_seen_at: string;
}

export interface AgentControlListResponse {
  agents: AgentControlState[];
}

// ──────────────────────────────────────────────────────────────────────
// Client
// ──────────────────────────────────────────────────────────────────────

const BASE = '/agent-control';

export const agentControlApi = {
  list: (): Promise<AgentControlListResponse> => api.request<AgentControlListResponse>(BASE),

  applyAction: (id: string, action: AgentControlAction): Promise<AgentControlState> =>
    api.request<AgentControlState>(`${BASE}/${id}/action`, {
      method: 'POST',
      body: JSON.stringify({ action }),
    }),
};
